File: 강의자료/섹션4/3. 멘토링 ChatGPT.js
function countMentoringPairs(testResults) {
  let count = 0;
  let pairs = []; // 멘토-멘티 쌍 저장
  let TestCount = testResults.length; // 시험 횟수 3번
  let StudentCount = testResults[0].length; // 학생 수 4명
  // 모든 (멘토, 멘티) 쌍을 탐색
  for (let mentor = 1; mentor <= StudentCount; mentor++) {
    for (let mentee = 1; mentee <= StudentCount; mentee++) {
      if (mentor === mentee) continue; // 같은 학생은 멘토-멘티 불가
      let validPair = true;

      // 모든 시험에서 멘토가 멘티보다 앞서는지 확인
      for (let test = 0; test < TestCount; test++) {
        let rank = testResults[test];
        if (rank.indexOf(mentor) > rank.indexOf(mentee)) {
          validPair = false;
          break;
        }
      }
      if (validPair) {
        count++;
        pairs.push([mentor, mentee]); // 멘토-멘티 쌍 추가
      }
    }
  }

  console.log("총 가능한 멘토-멘티 쌍 개수:", count);
  console.log("가능한 멘토-멘티 쌍:");
  pairs.forEach((pair) => console.log(`멘토: ${pair[0]}, 멘티: ${pair[1]}`));
}

const testResults = [
  [3, 4, 1, 2],
  [4, 3, 2, 1],
  [3, 1, 4, 2],
];

countMentoringPairs(testResults);

File: 강의자료/섹션4/3. 멘토링 ChatGPT.py
def count_mentoring_pairs(N, M, test_results):
    count = 0

    # 모든 (멘토, 멘티) 쌍을 탐색
    for mentor in range(1, N + 1):
        for mentee in range(1, N + 1):
            if mentor == mentee:
                continue  # 같은 학생은 멘토-멘티 불가

            valid_pair = True  # 모든 테스트에서 유효해야 함

            # 모든 시험에서 멘토가 멘티보다 높은 등수인지 확인
            for test in range(M):
                rank = test_results[test]
                if rank.index(mentor) > rank.index(mentee):
                    valid_pair = False
                    break
            
            if valid_pair:
                count += 1

    return count

# 예제 입력
N, M = 4, 3
test_results = [
    [3, 4, 1, 2],
    [4, 3, 2, 1],
    [3, 1, 4, 2]
]

print(count_mentoring_pairs(N, M, test_results))  # 출력: 3
